### PR TITLE
Hide interpreter unwinding behind a feature flag

### PR DESCRIPTION
**Pull request by kakkoyun**
Signed-off-by: Kemal Akkoyun <kakkoyun@gmail.com>

### Why?

Python and Ruby interpreter unwinding is still considered experimental at this stage. 
We want to release them as experimental first to guarantee the best UX.
We also plan to extend supported interpreter versions in the meanwhile.

### <samp>🤖 Generated by Copilot at c6ec75d</samp>

This pull request adds configuration options and flags to enable or disable interpreter unwinding features for Python and Ruby in the CPU profiler. It also updates the BPF program, the Go code, and the tests to respect these options and improve the performance and memory usage of the profiler.

### <samp>🤖 Generated by Copilot at c6ec75d</samp>

*  Add and use `Config` struct to pass configuration parameters to the BPF program and simplify function signatures ([link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-57833d176ed681886b67a7fa2e4dc2f68d5e00cb1db269e7854fb3dafb45f54eL764-R787), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL80-R122), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL108-R134), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL116-L127), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL136-R157), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL159-R173), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL231-R279), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL309-R347), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL454-R485), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL465-R495), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL510-R540), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL533-R563), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL599-R631), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL610-R646), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL632-R665), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL640-R671), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL717-R750), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL737-R770), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL861-R894), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-06263ba437ad3bf3713d34a6a03f5156a4297f8ce44ff8ea0942ec467e73ee45L39-R45), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-5d1089899c40725bc35d0c9852093988f1ff2e9fac1d17af76f57174c7244db1R302), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-5d1089899c40725bc35d0c9852093988f1ff2e9fac1d17af76f57174c7244db1L313-R327))
* Add and check boolean fields to `unwinder_config_t` struct in the BPF program to indicate whether Python and Ruby unwinding are enabled or not ([link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-d1aaaafbaebd6c404e2cdbc14c4bd95e684b61a791b57dc3322a75789cc74edfR97-R98), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-d1aaaafbaebd6c404e2cdbc14c4bd95e684b61a791b57dc3322a75789cc74edfL640-R660))
* Add and check boolean flags to the `FlagsHidden` struct and the `InfoManager` struct in the Go code to enable or disable the interpreter unwinding features ([link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-57833d176ed681886b67a7fa2e4dc2f68d5e00cb1db269e7854fb3dafb45f54eL219-R225), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-bc9c9211670f079eed672ee83ccef854706b1a639e4d9a0a28c92f5b4f5f0eb7R137-R138), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-bc9c9211670f079eed672ee83ccef854706b1a639e4d9a0a28c92f5b4f5f0eb7R152), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-bc9c9211670f079eed672ee83ccef854706b1a639e4d9a0a28c92f5b4f5f0eb7R184), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-bc9c9211670f079eed672ee83ccef854706b1a639e4d9a0a28c92f5b4f5f0eb7L265-R282))
* Modify the `loadBPFModules` function in the Go code to load and initialize the Python and Ruby unwinder modules only if the corresponding flags are enabled, and to add logging messages ([link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL231-R279), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL302-R311), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL334-R372), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dR378), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL369-R399))
* Modify the `symbols` map in the BPF program to have a maximum entry of 1 by default, to reduce the memory usage when the interpreter unwinding features are disabled ([link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-c0df7f364d90c13138234c063bff08f5aff8c088e0d1550ac462ce43c99e3085L56-R56))
* Add a logging message before attaching the perf events to all CPUs in the `Run` function in the Go code ([link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL704-R738))
* Rename the `loadBpfProgram` function to `loadBPFModules` in the Go code, to reflect that it loads multiple BPF modules for different unwinders ([link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL231-R279), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL632-R665), [link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-06263ba437ad3bf3713d34a6a03f5156a4297f8ce44ff8ea0942ec467e73ee45L39-R45))
* Modify the logging message in the `loadBPFModules` function in the Go code to use a lower-case letter for the first word, to be consistent with the other logging messages ([link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dL369-R399))
* Add an import statement for the `io/fs` package in the `cpu.go` file in the Go code, to use the `fs.File` interface in the `loadBPFModules` function ([link](https://github.com/parca-dev/parca-agent/pull/2017/files?diff=unified&w=0#diff-a3297cf4e3be27cabb198588463f4c04ca0f4aa5d81cbe206e467e1f1f7e772dR28))

### Test Plan

- Test on local system by using all combination of the flags while running Python and Ruby interpreters.

### CODE REVIEW

**Review by brancz — APPROVED** [review]

File: bpf/cpu/cpu.bpf.c [hunk 1]
```diff
@@ -94,6 +94,8 @@ struct unwinder_config_t {
   bool filter_processes;
   bool verbose_logging;
   bool mixed_stack_enabled;
+  bool python;
```
Comment (javierhonduco) [inline bpf/cpu/cpu.bpf.c:97]: nit: perhaps `python_enabled` and `ruby_enabled` could show the intent of these flags a bit better

File: bpf/cpu/shared.h [hunk 2]
```diff
@@ -53,7 +53,7 @@ struct {
 
 struct {
     __uint(type, BPF_MAP_TYPE_HASH);
-    __uint(max_entries, 64000);
+    __uint(max_entries, 1); // default: 64000. Set in the user-space.
```
Comment (javierhonduco) [inline bpf/cpu/shared.h:56]: nit: As the max entries value when enabled lives in userspace, should we remove the "default: 64000"? It's one more thing to keep in sync, a small one, though 😄

File: pkg/profiler/cpu/cpu.go [hunk 3]
```diff
@@ -77,54 +78,65 @@ const (
 	configKey                = "unwinder_config"
 )
 
-type Config struct {
+// UnwinderConfig gets sync to BPF module.
+type UnwinderConfig struct {
 	FilterProcesses   bool
 	VerboseLogging    bool
 	MixedStackWalking bool
+	Python            bool
```
Comment (javierhonduco) [inline pkg/profiler/cpu/cpu.go:86]: nit: same as in the BPF code 😄

File: pkg/profiler/cpu/cpu.go [hunk 4]
```diff
@@ -77,54 +78,65 @@ const (
...
+
+	DebugProcessNames []string
+
+	DWARFUnwindingDisabled         bool
+	DWARFUnwindingMixedModeEnabled bool
+	BPFVerboseLoggingEnabled       bool
+
+	PythonUnwindingEnabled bool
+	RubyUnwindingEnabled   bool
+}
```
Comment (javierhonduco) [inline pkg/profiler/cpu/cpu.go:108]: Thanks for separating the configs, it's way nicer now and long overdue to be honest 😄

File: pkg/profiler/cpu/maps.go [hunk 5]
```diff
@@ -1604,11 +1663,11 @@ func (m *bpfMaps) setUnwindTableForMapping(buf *profiler.EfficientBuffer, pid in
 			restChunks = restChunks[threshold:]
 
 			if currentChunk[0].IsEndOfFDEMarker() {
-				level.Error(m.logger).Log("msg", "First row of a chunk should not be a marker")
```
Comment (javierhonduco) [inline pkg/profiler/cpu/maps.go:1607]: wondering why the linter never complain about these ones

**Review by javierhonduco — APPROVED** [review]
LGTM, left some minor nits

File: pkg/profiler/cpu/maps.go (hunk 5 above)
Comment (kakkoyun) [reply pkg/profiler/cpu/maps.go:1607]: 🤷 We need to revisit our configuration.

**Pull request merged**